app/src/main/java/fr/gaulupeau/apps/Poche/data/QueueHelper.java

```diff
@@ -1,307 +1,325 @@
+            this.queueItems = queueItems;
+            this.negatedItems = negatedItems;
+            this.updated = updated;
 
-        // supposed to be called in a transaction, so separate operations are probably fine
-        for(QueueItem item: items) {
-            queueItemDao.delete(item);
+            this.completed = new ArrayList<>(this.queueItems.size());
         }
-    }
-
-    public boolean changeArticle(int articleID, ArticleChangeType articleChangeType) {
-        Log.d(TAG, String.format("changeArticle(%d, %s) started", articleID, articleChangeType));
 
-        QueueItem existingChangeItem = null;
+        public List<QueueItem> getQueueItems() {
+            return queueItems;
+        }
 
-        for(QueueItem item: getQueuedItemsForArticle(articleID)) {
-            switch(item.getAction()) {
-                case ARTICLE_CHANGE:
-                    existingChangeItem = item;
-                    Log.d(TAG, "changeArticle() found existing change item");
-                    break;
+        public void completed(QueueItem queueItem) {
+            updated.remove(queueItem);
+            completed.add(queueItem);
+        }
 
-                case ARTICLE_DELETE:
-                    Log.d(TAG, "changeArticle(): already in queue for Deleting; ignoring");
-                    return false;
-            }
+        public boolean hasChanges() {
+            return !completed.isEmpty() || !updated.isEmpty() || !negatedItems.isEmpty();
         }
 
-        boolean queueChanged = false;
+        public int itemsLeft() {
+            return queueItems.size() - completed.size();
+        }
 
-        if(existingChangeItem != null) {
-            Log.v(TAG, "changeArticle() existing changes: " + existingChangeItem.getExtra());
+    }
 
-            EnumSet<ArticleChangeType> changes
-                    = ArticleChangeType.stringToEnumSet(existingChangeItem.getExtra());
+    private static final String TAG = QueueHelper.class.getSimpleName();
 
-            if(!changes.contains(articleChangeType)) {
-                Log.d(TAG, "changeArticle() adding the change to change set");
+    private final DaoSession daoSession;
+    private final QueueItemDao queueItemDao;
 
-                changes.add(articleChangeType);
-                existingChangeItem.setExtra(ArticleChangeType.enumSetToString(changes));
+    public QueueHelper(DaoSession daoSession) {
+        this.daoSession = daoSession;
+        queueItemDao = daoSession.getQueueItemDao();
+    }
 
-                queueItemDao.update(existingChangeItem);
+    public QueueState getQueueState() {
+        List<QueueItem> queueItems = getQueueItems();
 
-                queueChanged = true;
-            } else {
-                Log.d(TAG, "changeArticle() change type is already queued: " + articleChangeType);
-            }
-        } else {
-            enqueueArticleChange(articleID, articleChangeType);
-            queueChanged = true;
+        if (queueItems.size() <= 1) {
```

Comment:
No, it means "if there is 0 or only 1 item, return as is" because there's nothing to optimize in these cases. Updated with `queueItems.size() < 2` - that should be more obvious.